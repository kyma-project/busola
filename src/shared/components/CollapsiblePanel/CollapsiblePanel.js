import React, { useState } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { LayoutPanel, Button } from 'fundamental-react';
import './CollapsiblePanel.scss';

export const CollapsiblePanel = ({
  children,
  title,
  className,
  actions,
  isOpenInitially = true,
}) => {
  const [isOpen, setIsOpen] = useState(isOpenInitially);

  const switchOpen = e => {
    e.stopPropagation();
    // ensure event didn't come from DOM propagation
    if (e.target === e.currentTarget) {
      setIsOpen(!isOpen);
    }
  };

  return (
    <LayoutPanel className={classNames('collapsible-panel', className)}>
      <LayoutPanel.Header onClick={switchOpen}>
        <h3 className="fd-panel__title" onClick={switchOpen}>
          {title}
        </h3>
        <LayoutPanel.Actions>
          {actions}
          <Button
            icon={isOpen ? 'navigation-up-arrow' : 'navigation-down-arrow'}
            design="Transparent"
            onClick={switchOpen}
          />
        </LayoutPanel.Actions>
      </LayoutPanel.Header>
      <LayoutPanel.Body
        className={isOpen ? 'body body--open' : 'body body--closed'}
      >
        {children}
      </LayoutPanel.Body>
    </LayoutPanel>
  );
};

CollapsiblePanel.propTypes = {
  children: PropTypes.element.isRequired,
  title: PropTypes.string.isRequired,
  className: PropTypes.string,
  actions: PropTypes.element,
  isOpenInitially: PropTypes.bool,
};
