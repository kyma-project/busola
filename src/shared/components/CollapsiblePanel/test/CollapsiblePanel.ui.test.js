import renderer from 'react-test-renderer';
import { LayoutPanel } from 'fundamental-react';
import { mount, shallow } from 'enzyme';

import { CollapsiblePanel } from 'shared/components/CollapsiblePanel/CollapsiblePanel';
import { Button } from '@ui5/webcomponents-react';

describe('Collapsible Panel', () => {
  const exceptPanelVisible = (component, isVisible) => {
    const body = component.find(LayoutPanel.Body);
    expect(body.hasClass('body--closed')).toBe(!isVisible);
    expect(body.hasClass('body--open')).toBe(isVisible);
  };

  it('Renders with minimal props (opened state)', () => {
    const component = renderer.create(
      <CollapsiblePanel
        children={<p>test</p>}
        title="Collapsible panel test"
      />,
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('Renders in closed state', () => {
    const component = renderer.create(
      <CollapsiblePanel
        children={<p>test</p>}
        title="Collapsible panel test"
        isOpenInitially={false}
      />,
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('Applies custom class names to Panel', () => {
    const component = shallow(
      <CollapsiblePanel
        children={<p>test</p>}
        title="Collapsible panel"
        className="test-class-name"
      />,
    );
    expect(component.find(LayoutPanel).hasClass('test-class-name')).toBe(true);
  });

  it('Is initially closed when isOpenInitially is false', () => {
    const component = shallow(
      <CollapsiblePanel
        children={<p>test</p>}
        title="Collapsible panel"
        isOpenInitially={false}
      />,
    );

    expect(component.find(LayoutPanel.Body).hasClass('body body--closed')).toBe(
      true,
    );
    expect(component.find(LayoutPanel.Body).hasClass('body body--open')).toBe(
      false,
    );
  });

  it('Opens and closes (chevron)', async () => {
    const component = mount(
      <CollapsiblePanel children={<p>test</p>} title="Collapsible panel" />,
    );
    const button = component.find(LayoutPanel.Header);

    // initially opened
    exceptPanelVisible(component, true);

    // close
    button.simulate('click');
    await component.update();
    exceptPanelVisible(component, false);

    // open again
    button.simulate('click');
    await component.update();
    exceptPanelVisible(component, true);
  });

  it('Opens and closes (header)', async () => {
    const component = mount(
      <CollapsiblePanel children={<p>test</p>} title="Collapsible panel" />,
    );
    const header = component.find(LayoutPanel.Header);

    // initially opened
    exceptPanelVisible(component, true);

    // close
    header.simulate('click');
    await component.update();
    exceptPanelVisible(component, false);

    // open again
    header.simulate('click');
    await component.update();
    exceptPanelVisible(component, true);
  });

  it('Does not open modal when clicking on custom actions', async () => {
    const child = <p>test</p>;
    const component = mount(
      <CollapsiblePanel
        children={child}
        title="Collapsible panel"
        actions={
          <Button id="action-button" onClick={() => {}}>
            test
          </Button>
        }
      />,
    );

    const button = component.find('ui5-button');

    // close
    button.simulate('click');
    await component.update();

    // expect it's still opened
    exceptPanelVisible(component, true);
  });
});
