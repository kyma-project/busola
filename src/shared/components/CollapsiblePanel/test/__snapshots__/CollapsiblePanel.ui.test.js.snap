// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Collapsible Panel Renders in closed state 1`] = `
<div
  className="fd-layout-panel collapsible-panel"
>
  <div
    className="fd-layout-panel__header"
    onClick={[Function]}
  >
    <h3
      className="fd-panel__title"
      onClick={[Function]}
    >
      Collapsible panel test
    </h3>
    <div
      className="fd-layout-panel__actions"
    >
      <button
        className="fd-button"
        design="Transparent"
        icon="navigation-down-arrow"
        onClick={[Function]}
        type="button"
      />
    </div>
  </div>
  <div
    className="fd-layout-panel__body body body--closed"
  >
    <p>
      test
    </p>
  </div>
</div>
`;

exports[`Collapsible Panel Renders with minimal props (opened state) 1`] = `
<div
  className="fd-layout-panel collapsible-panel"
>
  <div
    className="fd-layout-panel__header"
    onClick={[Function]}
  >
    <h3
      className="fd-panel__title"
      onClick={[Function]}
    >
      Collapsible panel test
    </h3>
    <div
      className="fd-layout-panel__actions"
    >
      <button
        className="fd-button"
        design="Transparent"
        icon="navigation-up-arrow"
        onClick={[Function]}
        type="button"
      />
    </div>
  </div>
  <div
    className="fd-layout-panel__body body body--open"
  >
    <p>
      test
    </p>
  </div>
</div>
`;
